Improved code style in physics units

jirikuncar opened the pull request:
```console
$ pip install pytest pytest-cov pytest-pycodestyle
$ pytest --pep8 --doctest-module --cov=sympy.physics.units --cov-report=term-missing ./sympy/physics/units/
```

Reformatting can be done by running following commands:
```console
$ pip install -e git+https://github.com/jirikuncar/yapf.git@sympy-style
$ yapf -ir sympy/physics/units/  # code style
$ isort -r sympy/physics/units/  # keep sorted imports
```

@asmeurer @moorepants @Upabjojr please let me know if you want to tweak any other formatting option.

Upabjojr commented:
I don't really like the new formatting. Sometimes it's clearer not to leave spaces around `*` and `/`.

jirikuncar commented:
@Upabjojr I am using default PEP8 recommendations.

jirikuncar commented:
```console
$ pytest --pdb --pep8 --doctest-module --cov=sympy.physics.units --cov-report=term-missing ./sympy/physics/units/
========================================== test session starts ==========================================
platform darwin -- Python 3.6.1, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
architecture: 64-bit
cache:        yes
ground types: python

rootdir: /Users/jirka/.polybox/src/sympy, inifile: setup.cfg
plugins: xdist-1.18.1, pycodestyle-1.0.6, cov-2.5.1
collected 83 items

sympy/physics/units/__init__.py ..
sympy/physics/units/definitions.py .
sympy/physics/units/dimensions.py ..
sympy/physics/units/prefixes.py ..
sympy/physics/units/quantities.py ..
sympy/physics/units/unitsystem.py .
sympy/physics/units/util.py ..
sympy/physics/units/systems/SI.py ..
sympy/physics/units/systems/__init__.py .
sympy/physics/units/systems/mks.py .
sympy/physics/units/systems/mksa.py .
sympy/physics/units/systems/natural.py .
sympy/physics/units/tests/__init__.py .
sympy/physics/units/tests/test_dimensions.py ........
sympy/physics/units/tests/test_dimensionsystem.py ................
sympy/physics/units/tests/test_prefixes.py .....
sympy/physics/units/tests/test_quantities.py ..................x
sympy/physics/units/tests/test_unitsystem.py ........
sympy/physics/units/tests/test_util.py ........

---------- coverage: platform darwin, python 3.6.1-final-0 -----------
Name                                                Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------
sympy/physics/units/__init__.py                        35      0   100%
sympy/physics/units/definitions.py                    102      0   100%
sympy/physics/units/dimensions.py                     231      0   100%
sympy/physics/units/prefixes.py                        91      0   100%
sympy/physics/units/quantities.py                      95      0   100%
sympy/physics/units/systems/SI.py                      15      0   100%
sympy/physics/units/systems/__init__.py                 4      0   100%
sympy/physics/units/systems/mks.py                     15      0   100%
sympy/physics/units/systems/mksa.py                    14      0   100%
sympy/physics/units/systems/natural.py                 10      0   100%
sympy/physics/units/tests/__init__.py                   0      0   100%
sympy/physics/units/tests/test_dimensions.py           55      0   100%
sympy/physics/units/tests/test_dimensionsystem.py      73      0   100%
sympy/physics/units/tests/test_prefixes.py             32      0   100%
sympy/physics/units/tests/test_quantities.py          163      0   100%
sympy/physics/units/tests/test_unitsystem.py           40      0   100%
sympy/physics/units/tests/test_util.py                 65      0   100%
sympy/physics/units/unitsystem.py                      41      0   100%
sympy/physics/units/util.py                            45      0   100%
---------------------------------------------------------------------------------
TOTAL                                                1126      0   100%


================================= 82 passed, 1 xfailed in 3.53 seconds ==================================
```

Upabjojr commented:
> I am using default PEP8 recommendations.

We try to follow them, but sometimes they are not ideal.

moorepants commented:
@jirikuncar Note that we are not strict adherents to pep8, especially on a few things like variable names for symbols and spaces around multiply/divide.

moorepants commented on .editorconfig:
```diff
@@ -0,0 +1,25 @@
+root = true
+
+[*]
+indent_style = space
```
note sure this would apply to all files, for example the sphinx Makefile in doc.

moorepants commented on .editorconfig:
```diff
@@ -0,0 +1,25 @@
+
+# RST files (used by sphinx)
+[*.rst]
+indent_size = 4
```
This should be 3 I believe.

moorepants commented on .editorconfig:
```diff
@@ -0,0 +1,25 @@
+charset = utf-8
+
+# Python files
+[*.py]
```
Is there a linewrap setting?

jirikuncar replied:
`max_line_length`

moorepants commented on setup.cfg:
```diff
@@ -0,0 +1,8 @@
+[pycodestyle]
```
What is a "setup.cfg" file for? Is it tied to a specific program?

jirikuncar replied:
It's a common file used by setuptools and many Python related utilities.

asmeurer replied:
I think we shouldn't add this file. People can have their own configurations for whatever code quality tools they use.

moorepants commented on sympy/physics/units/__init__.py:
```diff
@@ -28,12 +27,13 @@
     expression expressed in another unit.
 
 """
+# isort:skip_file
```
I'm not sure people will know what these are and I think it would follow convention more to put it at the top just below the `# _*_ coding: utf-8 _*_`. And note that it is related to pytest (or whatever it is).

jirikuncar replied:
It's related only to `isort` utility which helps with keeping the imports well organized. I can move it to the top if you prefer.

moorepants commented on sympy/physics/units/definitions.py:
```diff
@@ -83,87 +100,133 @@
+    "gravitational_constant",
+    length**3 * mass**-1 * time**-2,
+    6.67408e-11 * m**3 / (kg * s**2),
+    abbrev="G")
```
Where does this indentation convention come from? I personally prefer leaving line 135 on 134 and then indenting the next lines to the edge of the open parentheses.

jirikuncar replied:
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

Upabjojr replied:
I like this restyling, except for the spaces between `*` and `/`.

moorepants commented on sympy/physics/units/definitions.py:
```diff
@@ -83,87 +100,133 @@
-planck_temperature = Quantity("planck_temperature", temperature, sqrt(hbar*speed_of_light**5/G/boltzmann**2), "T_P")
-planck_length = Quantity("planck_length", length, sqrt(hbar*G/speed_of_light**3), "l_P")
+planck_mass = Quantity("planck_mass", mass,
+                       sqrt(hbar * speed_of_light / G), "m_P")
```
This is my preferred way. I'm not sure how consistent sympy's files are. Is there a reason to have both in this file?

jirikuncar replied:
It's done by a [YAPF](https://github.com/google/yapf) tool. I can play a bit with weights to prefer one way, but sometimes it is not possible since the expression is too long.

moorepants commented on sympy/physics/units/dimensions.py:
```diff
@@ -149,36 +153,43 @@ def __add__(self, other):
+        return Dimension(self.name * other.name)
 
     def __div__(self, other):
+        """Division works only between two dimensions."""
```
There is no need to add docstrings to dunder methods, in general. We don't require that and you'll find that the large majority do not have them.

jirikuncar replied:
I think it is not bad to have them anyway.

moorepants commented on sympy/physics/units/dimensions.py:
```diff
@@ -414,126 +417,107 @@ def get_dim(self, dim):
         return matrix
 
-    #@cacheit
+    # @cacheit
```
Just remove that. Or turn into a useful note: `NOTE : This may be able to be cached.`.

moorepants commented on sympy/physics/units/tests/test_prefixes.py:
```diff
@@ -1,22 +1,32 @@
 
-def test_prefix_unit():
-    from sympy.physics.units import Quantity, Dimension
+    m = Quantity("meter", 1, 6)
```
This was actually a test it seems. Why did you change it?

jirikuncar replied:
I have just added few more lines. The test `test_prefix_unit` is untouched.

moorepants commented on sympy/physics/units/tests/test_util.py:
```diff
@@ -2,30 +2,13 @@
+    hbar, hour, inch, joule, kelvin, kilogram, kilometer, length, meter, \
+    mile, minute, newton, planck, planck_length, planck_mass, \
+    planck_temperature, planck_time, radians, second, speed_of_light, \
+    steradian, time
```
We prefer using parentheses to group multi-line imports or multiple import statements to backslashes.

jirikuncar replied:
I can change that in `.editorconfig`

moorepants commented on sympy/physics/units/unitsystem.py:
```diff
@@ -24,16 +21,15 @@ class UnitSystem(object):
-        if self.is_consistent is False:
-            raise ValueError("The system with basis '%s' is not consistent"
-                             % str(self._base_units))
+        assert self.is_consistent  # test is performed in DimensionSystem
```
This doesn't look quite equivalent. Maybe the error message was important to display in that way.

jirikuncar replied:
`# test is performed in DimensionSystem` there is no way you can reach the old message.

moorepants commented on sympy/physics/units/unitsystem.py:
```diff
@@ -46,55 +42,38 @@ def __init__(self, base, units=(), name="", descr=""):
-            raise KeyError(key)
-
-        return u
-
```
Why remove?

moorepants commented on sympy/physics/units/util.py:
```diff
@@ -25,60 +21,54 @@ def dim_simplify(expr):
-
-    raise ValueError("Cannot be simplifed: %s", expr)
+    _, expr = Quantity._collect_factor_and_dimension(expr)
+    return expr
```
What is this change?

jirikuncar replied:
The code is the same in `_collect_factor_and_dimension`.

moorepants commented on sympy/physics/units/util.py:
```diff
@@ -25,60 +21,54 @@ def dim_simplify(expr):
+        i
+        for x in target_dims
+        for i in x.get_dimensional_dependencies(mark_dimensionless=True)
+    }
```
Odd looking convention. I think it is better to keep them on a line and break once after target_dims, as comprehensions are supposed to be oneliners in some sense.

jirikuncar replied:
There is a readability issue if you don't break the line since `i` is not used on `for x in target_dims`.

moorepants commented:
Nice to see this pass with pytest and the extra linters. I think, in general, PRs like this will be favored, especially if we can move to a modern pytest for testing. Here are a few pointers:

1. Setup pep8 checker to not worry about spaces around `*` and `/`. You'll get complaints about trying to change this broadly. Spaces around `+` and `-` are probably ok. We allow either convention for the multiple/divide.
2. I'm not sure we care about the import sorting unless it can fix circular import dependencies. If it can, then I'd include this, if not I would not try sorting the imports.
3. You've got at least three multi-line indent styles in these files. We favor parentheses to backslashes and I'm not sure what the most common way to indent is. Might be worth looking through different modules.
4. In general, don't try to pep8tify all the files in sympy (if that is your plan). We are not strict pep8 adherents and you'll get resistance if you try to enforce it strictly in the codebase. There are only a handful of pep8 guidelines that won't get that resistance.
5. If you goal is to get pytest running on the modules, I'd submit a pull request with only changes for that. It'd be better if your linting related pull requests were separate PRs.

If you can address my comments above this is mergable.

jirikuncar commented:
> Setup pep8 checker to not worry about spaces around * and /. You'll get complaints about trying to change this broadly. Spaces around + and - are probably ok. We allow either convention for the multiple/divide.

You can use tools like YAPF or autopep8 which will reformat everything automatically. I can have a look if we can set the space around `*` and `/` to be optional.

> I'm not sure we care about the import sorting unless it can fix circular import dependencies. If it can, then I'd include this, if not I would not try sorting the imports.

If you start module by module, it is possible to achieve it. I can slowly fix everything in `physics` module if it helps.

> You've got at least three multi-line indent styles in these files. We favor parentheses to backslashes and I'm not sure what the most common way to indent is. Might be worth looking through different modules.

I have changed it to `Hanging Grid` style (4 - https://github.com/timothycrosley/isort#multi-line-output-modes).

> In general, don't try to pep8tify all the files in sympy (if that is your plan). We are not strict pep8 adherents and you'll get resistance if you try to enforce it strictly in the codebase. There are only a handful of pep8 guidelines that won't get that resistance.

There were not so many problems in `sympy.physics.units` module, but I understand that changing the whole code base could bring more problems then benefits. I would just hope that all new code will follow some style guidelines.

> If you goal is to get pytest running on the modules, I'd submit a pull request with only changes for that. It'd be better if your linting related pull requests were separate PRs.

Yes the goal is to have `pytest` running everywhere. I hope you can accept it as is since I have addressed most of your comments.

Upabjojr commented:
I believe the file `.editorconfig` should be moved to another PR. It's unrelated to the units and needs a separate discussion.

I don't like the spaces between multiplication and division signs. I think that PEP8 is not ideal for our case (PEP8 was probably not thought for large mathematical expressions).

Upabjojr commented:
Same for `setup.cfg`... not in this PR please.

jirikuncar commented:
> I believe the file `.editorconfig` should be moved to another PR. It's unrelated to the units and needs a separate discussion.

Sure, it has been merged in #12934.

jirikuncar commented:
> Same for `setup.cfg`... not in this PR please.

`setup.cfg` is related because it stores default options for tools used in this PR.

jirikuncar commented:
> I don't like the spaces between multiplication and division signs. I think that PEP8 is not ideal for our case (PEP8 was probably not thought for large mathematical expressions).

It is a matter of personal taste and the recommendation has been discussed in wider Python community. There was some consensus while making PEP8 and if you ask different people from 1967 people that have forked this repo, you might get different answers. Hence, I would recommend to use "standard" configuration for Python tools to avoid repeating discussion later on.

moorepants commented:
> It is a matter of personal taste and the recommendation has been discussed in wider Python community.

@jirikuncar I think you may be misunderstanding @Upabjojr and me here. We all understand that there are style conventions that have been adopted by much of the wider Python community but SymPy does not adhere to all of them. In particular, we do not enforce spaces around * and / and we are not interested in merging PRs that change existing code in the SymPy codebase wrt to this detail. We allow with or without spaces in new commits, but do not undo this or want to make the codebase uniform in that respect.

If you want to get an idea of the SymPy "philosophy" around pep8 and linting issues I recommend searching the mailing list and the repo for "pep8" to see similar conversations, as we are having here.

https://github.com/sympy/sympy/search?q=pep8&type=Issues&utf8=%E2%9C%93

https://groups.google.com/forum/#!searchin/sympy/pep8%7Csort:relevance

Here is my recommendation for getting PRs merged:

1. Create PRs that pass pytest without any linter plugins. These also shouldn't break our current testing suite. These will be quickly merged and will be a great contribution to getting us off our old fork.
2. Create PRs that are strictly about linting. Keep each style guideline in separate prs: pep8, yapf, isort, etc. These should not be super strict and adhere to **SymPy's** style. The easiest way to get these merged is to do the most basic pep8 changes. We do not lint in our CI service because we do not want to impose a style rule set. SymPy only loosely follows pep8, for example, and we treat it as a guideline not a rule.
3. Create separate PRs for introducing these various configuration files.

jirikuncar commented:
@moorepants I have read a small sample of the discussions and I don't have a good feeling from it. I hope there will be someone who will recognise benefits of automatically formatted code and strict code/doc style requirements.

I would like to point out the sorted imports should be must no matter how long it takes to enforce other code style rules.

moorepants commented:
> I hope there will be someone who will recognise benefits of automatically formatted code and strict code/doc style requirements.

I've worked with SymPy for over 8 years and I think this is wishful thinking. There are people that lean this way, but others that do not.  I, for one, don't mind picking a set of style rules and using them but we've never been able to come to consensus on that in the past. My guess is that you can get most things merged but I recommend breaking things up into different types of pull requests, otherwise you will unfortunately find that these style PRs will get bogged down. If you break them up I bet you can get the majority of them in with less resistance.

asmeurer commented:
This is what PEP 8 says:

> If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator.
>
> Yes:
>
>     i = i + 1
>     submitted += 1
>     x = x*2 - 1
>     hypot2 = x*x + y*y
>     c = (a+b) * (a-b)
>
> No:
>
>     i=i+1
>     submitted +=1
>     x = x * 2 - 1
>     hypot2 = x * x + y * y
>     c = (a + b) * (a - b)

So not having spaces around multiplication *is* following PEP 8. 

To the broader discussion, I'm against aggressive linting tests, because there are too many false positives, which cause unnecessary annoyance.

asmeurer commented on sympy/physics/units/__init__.py:
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-
+# isort:skip_file
```
What is isort?

moorepants commented:
Funny about this particular one, it changed in 2012: https://stackoverflow.com/a/37956466/467314

asmeurer commented on sympy/physics/units/prefixes.py:
```diff
@@ -183,11 +195,11 @@ def prefix_unit(unit, prefixes):
-gibi = Prefix('gibi', 'Y', 30, 2),
-tebi = Prefix('tebi', 'Y', 40, 2),
-pebi = Prefix('pebi', 'Y', 50, 2),
+kibi = Prefix('kibi', 'Y', 10, 2)
```
Unlike the other changes in this commit, this appears to be a legitimate bug.

asmeurer commented:
Yes, that's right, because the core Python devs realized that PEP 8 was wrong there (yes, PEP 8 is capable of being wrong). SymPy has always used the convention of spaces around + and - and no spaces around other binary operators, because it's pretty obvious once you have any nontrivial expression (like `3 * x ** 2 - 2 * x - 3`) that it's impossible to read without proper grouping. 

I personally think PEP 8 is wrong in other ways, e.g., the rule on hanging indents (and I'm [not the only one](https://twitter.com/asmeurer/status/800465167544791040)). And apparently so many people disagreed about the line length thing that there is some language in there now about longer lines being OK.

For SymPy, our rule has always been to keep things readable, but for the most part, everyone is free to use their own coding style. We have a couple of code quality tests like trailing whitespace (which mainly prevents unnecessary diffs). 

I've said before, I would much (*much* **much**) rather see pull requests that fix to actual logical code errors than style "errors". pyflakes currently reports over 3000 errors on sympy/. A lot of them are harmless things like unused imports, but some are actual bugs (for those of you who don't know, pyflakes, unlike most other Python linters, doesn't warn about style issues, and tries to avoid false positives). If we could fix all pyflakes bugs and actually run it in our test suite, that would be so *so* much better than if we could make our test suite force imports to be sorted and lines not being longer than 80 characters. One avoids actual bugs, the other does nothing more than add a marginal improvement to readability. 

I'm fine with this PR, after the operator space changes are reverted, although it would be much easier to review if the code quality fixes were separated into separate commits from the actual code changes, or better, a separate PR.

asmeurer commented:
By the way, our test suite should already work with pytest. If there are tests that don't run with it, we should fix it.

asmeurer commented:
> I would like to point out the sorted imports should be must no matter how long it takes to enforce other code style rules.

I think you have a typo here, but assuming you are saying that sorted imports are the most important style point, can you explain why? To me the order of imports is completely irrelevant.

moorepants commented:
> For SymPy, our rule has always been to keep things readable, but for the most part, everyone is free to use their own coding style.

We need to have some form of this statement in our developer docs (if not already there).

jirikuncar commented:
PEP8: we can ignore certain checks in `setup.cfg`. I trust your judgement about spaces around binary operators, but then it should be consistent and checked.
ISort: It's not only about sorting but also checking that there are no duplicates even by a mistake done by a developer. Moreover it usually reveal problems with circular imports which are caused mainly by wrong design decision.

Please have a look and merge the non controversial PRs I have extracted from this one. Thanks!

jirikuncar commented:
I hope that once the support for configurable spaces around binary operators gets to [YAPF](https://github.com/google/yapf/pull/427), we might be able to agree on common style.

Upabjojr commented:
@jirikuncar this PR is too long. Some concepts are already OK to merge, while some others aren't. Maybe you should split this PR in various smaller PRs.

jirikuncar commented:
@Upabjojr it has been split long time ago. I am not sure if I continue with this PR since there is no agreement on the code style.

jirikuncar committed:
Unified code style in physics units

Uses YAPF tool to keep unified formatting of all files.
One can run following commands in the given order to reproduce
the same code.

  $ yapf -ir sympy/physics/units/  # code style
  $ isort -r sympy/physics/units/  # keep sorted imports

jirikuncar commented:
Please note that this PR contains #13006 which can be merged separately.

jirikuncar commented:
I don't really get what is going on Travis with the randomly failing test. I hope a restart will fix it.

gxyd commented:
Are there still commits in this PR, that can be used? Did #13006 cherry-picked all the necessary commits?

gxyd commented:
@jirikuncar @asmeurer should I close this PR? Probably looks like the other PR did cherry-picked the important commits.

gxyd commented:
I am closing this. Seems like this important commits from this PR were already merged.
